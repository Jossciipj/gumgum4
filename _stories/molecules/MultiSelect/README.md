The `<MultiSelect>` component can be used when you need a select component to support multiple options simultaneously.

The state on the multi-select is handed by the parent component, the callback passed into should also take care of updating the dropdown options.

**NOTE**: This is a controlled component, so don't forget to pass the updated props to `MultiSelect` to see the changes.

**Example**:
```
state {
    options: [ { name: 'Some Name', value: 'some_value', selected: false}, ... ]
}
const handleMultiselectChange = (index, value, selected) => {
    // Update your options here
    const newOptions = ...;
    this.setState({ options: newOptions });
};
return(
    <MultiSelect
       placeholder="Select an option"
       options={ this.state.options }
       callback= { this.handleMultiselectChange }
    />
)
```

*Options format*:
```
[
    {
        name: 'Option 1',
        value: 'option_1',
        selected: false
    },
    {
        name: 'Option 2',
        value: 'option_2',
        selected: true
    },
    ...
]
```

The callback receives 3 paramenters:
- `index` {Number} position on the array of the option changed
- `value` {String} of the option change
- `selected`{Boolean} new value of the option. \n


Therefore, your callback should look like this:
```
function callback(index, value, selected){
     ...
}
```

If you want to add an option that "checks" all the other ones. You can create a new option with a value of `SELECT_ALL`, and in your callback handler you can check for that option:
```
if (value === 'SELECT_ALL' && selected) {
    // Checks all the options as true
    return someOptions.map( o => ({ ...o, selected: true }));
}
```
