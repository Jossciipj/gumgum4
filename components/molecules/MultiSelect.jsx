import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Checkbox from './Checkbox';
import cx from 'classnames';
import generateUID from '../utils/generateUID';

class MultiSelect extends Component {
    uid = generateUID(this);

    state = {
        isOpen: false
    };

    componentDidMount() {
        window.addEventListener('click', this._closeOnClickOutside);
    }

    componentWillUnmount() {
        window.removeEventListener('click', this._closeOnClickOutside);
    }

    _closeOnClickOutside = ({ target }) => {
        const el = this.container;
        if (!el.contains(target)) {
            this.setState({ isOpen: false });
        }
    };

    _toggleDropdown = () => this.setState(prevState => ({ isOpen: !prevState.isOpen }));

    _getPlaceholder = () => {
        const { placeholder, options } = this.props;
        const selectedOptions = options.filter(x => x.selected);
        return selectedOptions.map(x => x.name).join(', ') || placeholder;
    };

    render() {
        const { options, callback, placeholder, size, className, ...otherProps } = this.props;
        const { isOpen } = this.state;

        const isSmall = size === 'sm';
        const isExtraSmall = size === 'xs';

        const rootClass = cx('gds-multi-select', className, {
            'gds-multi-select--sm': isSmall,
            'gds-multi-select--xs': isExtraSmall,
            'gds-button-dropdown--active': isOpen
        });

        const btnClass = cx('gds-multi-select__button', {
            'gds-multi-select__button--sm': isSmall,
            'gds-multi-select__button--xs': isExtraSmall
        });

        const regionId = `MultiSelect_region_${this.uid}`;
        const labelId = `MultiSelect_label_${this.uid}`;

        return (
            <div className={rootClass} {...otherProps} ref={ref => (this.container = ref)}>
                <button
                    aria-expanded={isOpen}
                    aria-pressed={isOpen}
                    aria-controls={regionId}
                    tabIndex={0}
                    className={btnClass}
                    id={labelId}
                    type="button"
                    onClick={this._toggleDropdown}>
                    <div className="-ellipsis">{this._getPlaceholder()}</div>
                </button>
                <ul
                    aria-labelledby={labelId}
                    aria-hidden={!isOpen}
                    className="gds-multi-select__menu"
                    id={regionId}
                    role="region">
                    {options.map(({ name, value, selected }, index) => (
                        <li key={`item-${index}`} className="gds-multi-select__menu-item">
                            <div className="gds-multi-select__menu-link">
                                <div className="gds-form-group gds-multi-select__option">
                                    <Checkbox
                                        label={name}
                                        checked={selected}
                                        onChange={event => {
                                            callback(index, value, !selected);
                                        }}
                                    />
                                </div>
                            </div>
                        </li>
                    ))}
                </ul>
            </div>
        );
    }
}

MultiSelect.displayName = 'MultiSelect';

MultiSelect.propTypes = {
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    callback: PropTypes.func.isRequired,
    placeholder: PropTypes.string,
    size: PropTypes.oneOf(['xs', 'sm', '']),
    className: PropTypes.string
};

MultiSelect.defaultProps = {
    placeholder: '',
    className: ''
};

export default MultiSelect;
