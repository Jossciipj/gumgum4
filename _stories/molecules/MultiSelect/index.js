import React, { Component } from 'react';
import { text } from '@storybook/addon-knobs';
import { optionalSelect } from '../../../components/utils/optionalSelect';
import { action } from '@storybook/addon-actions';

import readme from './README.md';
import MultiSelect from '../../../components/molecules/MultiSelect';

const options = Array(10)
    .fill({})
    .map((o, i) => ({
        name: `Option ${i + 1}`,
        value: i,
        selected: false
    }));

const sizeOptions = {
    sm: 'sm',
    xs: 'xs',
    'No Value': ''
};

class MultiSelectOptions extends Component {
    state = {
        options
    };

    handleCallback = (index, value, selected) => {
        const options = this.state.options.map((o, i) => ({
            ...o,
            selected: i === index ? selected : o.selected
        }));

        this.setState({
            options
        });
    };

    render() {
        return (
            <MultiSelect
                placeholder={text('Placeholder', 'Select an option')}
                size={optionalSelect('Size', sizeOptions, '')}
                callback={this.handleCallback}
                options={this.state.options}
                className={text('className', '')}
            />
        );
    }
}

const component = () => <MultiSelectOptions />;

export default [readme, component];
