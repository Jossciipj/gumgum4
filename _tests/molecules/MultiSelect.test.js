/* globals mount */
import React from 'react';
import MultiSelect from '../../components/molecules/MultiSelect';

const defaultProps = {
    options: [
        {
            name: 'Name',
            value: 10,
            selected: true
        }
    ],
    callback: () => {},
    placeholder: 'placeholder',
    size: 'xs',
    className: 'custom-class'
};

describe('Expect <MultiSelect>', () => {
    it('to render', () => {
        const wrapper = mount(<MultiSelect {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });
});
