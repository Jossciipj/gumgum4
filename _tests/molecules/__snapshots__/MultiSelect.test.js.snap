// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <MultiSelect> to render 1`] = `
<MultiSelect
  callback={[Function]}
  className="custom-class"
  options={
    Array [
      Object {
        "name": "Name",
        "selected": true,
        "value": 10,
      },
    ]
  }
  placeholder="placeholder"
  size="xs"
>
  <div
    className="gds-multi-select custom-class gds-multi-select--xs"
  >
    <button
      aria-controls="MultiSelect_region_TXVsdGlTZWxlY3Q=0"
      aria-expanded={false}
      aria-pressed={false}
      className="gds-multi-select__button gds-multi-select__button--xs"
      id="MultiSelect_label_TXVsdGlTZWxlY3Q=0"
      onClick={[Function]}
      tabIndex={0}
      type="button"
    >
      <div
        className="-ellipsis"
      >
        Name
      </div>
    </button>
    <ul
      aria-hidden={true}
      aria-labelledby="MultiSelect_label_TXVsdGlTZWxlY3Q=0"
      className="gds-multi-select__menu"
      id="MultiSelect_region_TXVsdGlTZWxlY3Q=0"
      role="region"
    >
      <li
        className="gds-multi-select__menu-item"
        key="item-0"
      >
        <div
          className="gds-multi-select__menu-link"
        >
          <div
            className="gds-form-group gds-multi-select__option"
          >
            <Checkbox
              checked={true}
              label="Name"
              onChange={[Function]}
            >
              <div
                className="gds-form-group__checkbox"
              >
                <label
                  className="gds-form-group__checkbox-label"
                >
                  <input
                    checked={true}
                    className="gds-form-group__checkbox-input"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  <span
                    className="gds-form-group__checkbox-indicator"
                  />
                  Name
                </label>
              </div>
            </Checkbox>
          </div>
        </div>
      </li>
    </ul>
  </div>
</MultiSelect>
`;
